File: sdk/typescript/src/execution/output.ts
/**
 * Output utilities for the execution framework.
 *
 * Provides functions for truncating large outputs, detecting binary content,
 * parsing grep output, and stripping ANSI escape codes.
 */

import type { GrepMatch } from './types.js';

/** Default maximum output characters */
const DEFAULT_MAX_CHARS = 100_000;

/** Head portion of truncated output (20% of max) */
const HEAD_RATIO = 0.2;

/**
 * Truncate output that exceeds the maximum character limit.
 *
 * Keeps the first 20K characters (head) and last 80K characters (tail)
 * of a 100K max, with a truncation message in between.
 */
export function truncateOutput(
  output: string,
  maxChars?: number
): { text: string; truncated: boolean } {
  const max = maxChars ?? DEFAULT_MAX_CHARS;
  if (output.length <= max) {
    return { text: output, truncated: false };
  }

  const headSize = Math.floor(max * HEAD_RATIO);
  const tailSize = max - headSize;
  const omitted = output.length - headSize - tailSize;

  const head = output.slice(0, headSize);
  const tail = output.slice(-tailSize);
  const text = `${head}\n\n--- truncated ${String(omitted)} characters ---\n\n${tail}`;

  return { text, truncated: true };
}

/**
 * Detect binary content by checking for null bytes in the first 8KB.
 */
export function isBinary(buffer: Buffer): boolean {
  const checkLength = Math.min(buffer.length, 8192);
  for (let i = 0; i < checkLength; i++) {
    if (buffer[i] === 0) {
      return true;
    }
  }
  return false;
}

/**
 * Parse `grep -rn` output into structured GrepMatch objects.
 *
 * Expected format: `filepath:linenum:matched text`
 */
export function parseGrepOutput(output: string): GrepMatch[] {
  if (!output.trim()) {
    return [];
  }

  const matches: GrepMatch[] = [];
  const lines = output.split('\n');

  for (const line of lines) {
    if (!line) continue;

    // Match format: path:line:text
    const match = /^(.+?):(\d+):(.*)$/.exec(line);
    if (match?.[1] && match[2] && match[3] !== undefined) {
      matches.push({ path: match[1], line: parseInt(match[2], 10), text: match[3] });
    }
  }

  return matches;
}

/**
 * Strip ANSI escape codes from text.
 */
export function stripAnsi(text: string): string {
  // Matches all ANSI escape sequences
  // eslint-disable-next-line no-control-regex
  return text.replace(/\x1b\[[0-9;]*[a-zA-Z]/g, '');
}

File: sdk/python/polos/execution/output.py
"""Output utilities for the execution framework.

Provides functions for truncating large outputs, detecting binary content,
parsing grep output, and stripping ANSI escape codes.
"""

from __future__ import annotations

import re

from .types import GrepMatch

# Default maximum output characters
DEFAULT_MAX_CHARS = 100_000

# Head portion of truncated output (20% of max)
HEAD_RATIO = 0.2


def truncate_output(output: str, max_chars: int | None = None) -> tuple[str, bool]:
    """Truncate output that exceeds the maximum character limit.

    Keeps the first 20% characters (head) and last 80% characters (tail)
    of the max, with a truncation message in between.

    Args:
        output: The output string to potentially truncate.
        max_chars: Maximum character limit (default: 100,000).

    Returns:
        A tuple of (text, truncated) where truncated is True if output was truncated.
    """
    max_c = max_chars if max_chars is not None else DEFAULT_MAX_CHARS
    if len(output) <= max_c:
        return output, False

    head_size = int(max_c * HEAD_RATIO)
    tail_size = max_c - head_size
    omitted = len(output) - head_size - tail_size

    head = output[:head_size]
    tail = output[-tail_size:]
    text = f"{head}\n\n--- truncated {omitted} characters ---\n\n{tail}"

    return text, True


def is_binary(data: bytes) -> bool:
    """Detect binary content by checking for null bytes in the first 8KB.

    Args:
        data: The raw bytes to check.

    Returns:
        True if binary content is detected.
    """
    check_length = min(len(data), 8192)
    return any(data[i] == 0 for i in range(check_length))


def parse_grep_output(output: str) -> list[GrepMatch]:
    """Parse ``grep -rn`` output into structured GrepMatch objects.

    Expected format: ``filepath:linenum:matched text``

    Args:
        output: Raw grep output string.

    Returns:
        List of GrepMatch objects.
    """
    if not output.strip():
        return []

    matches: list[GrepMatch] = []
    pattern = re.compile(r"^(.+?):(\d+):(.*)$")

    for line in output.split("\n"):
        if not line:
            continue

        m = pattern.match(line)
        if m:
            matches.append(
                GrepMatch(
                    path=m.group(1),
                    line=int(m.group(2)),
                    text=m.group(3),
                )
            )

    return matches


def strip_ansi(text: str) -> str:
    """Strip ANSI escape codes from text.

    Args:
        text: Text potentially containing ANSI escape sequences.

    Returns:
        Cleaned text without ANSI codes.
    """
    return re.sub(r"\x1b\[[0-9;]*[a-zA-Z]", "", text)
